qiskit/chemistry/ground_state_calculation/adapt_vqe.py

```diff
@@ -0,0 +1,275 @@
+
+        Returns:
+            List of pairs consisting of gradient and excitation operator.
+        """
+        res = []
+        # compute gradients for all excitation in operator pool
+        for exc in excitation_pool:
+            # push next excitation to variational form
+            var_form.push_hopping_operator(exc)
+            # construct auxiliary VQE instance
+            vqe = self._solver.get_solver(self._transformation)
+            vqe.operator = operator
+            vqe.var_form = var_form
+            vqe.initial_point = var_form.preferred_init_points
+            # evaluate energies
+            parameter_sets = theta + [-self._delta] + theta + [self._delta]
+            energy_results = vqe._energy_evaluation(np.asarray(parameter_sets))
+            # compute gradient
+            gradient = (energy_results[0] - energy_results[1]) / (2 * self._delta)
+            res.append((np.abs(gradient), exc))
+            # pop excitation from variational form
+            var_form.pop_hopping_operator()
+
+        return res
+
+    @staticmethod
+    def _check_cyclicity(indices: List[int]) -> bool:
+        """
+        Auxiliary function to check for cycles in the indices of the selected excitations.
+
+        Args:
+            indices: the list of chosen gradient indices.
+        Returns:
+            Whether repeating sequences of indices have been detected.
+        """
+        cycle_regex = re.compile(r"(\b.+ .+\b)( \b\1\b)+")
+        # reg-ex explanation:
+        # 1. (\b.+ .+\b) will match at least two numbers and try to match as many as possible. The
+        #    word boundaries in the beginning and end ensure that now numbers are split into digits.
+        # 2. the match of this part is placed into capture group 1
+        # 3. ( \b\1\b)+ will match a space followed by the contents of capture group 1 (again
+        #    delimited by word boundaries to avoid separation into digits).
+        # -> this results in any sequence of at least two numbers being detected
+        match = cycle_regex.search(' '.join(map(str, indices)))
+        logger.debug('Cycle detected: %s', match)
+        # Additionally we also need to check whether the last two numbers are identical, because the
+        # reg-ex above will only find cycles of at least two consecutive numbers.
+        # It is sufficient to assert that the last two numbers are different due to the iterative
+        # nature of the algorithm.
+        return match is not None or (len(indices) > 1 and indices[-2] == indices[-1])
+
+    def compute_groundstate(self, driver: BaseDriver) -> 'AdaptVQEResult':
+        """Computes the ground state.
+
+        Args:
+            driver: a chemistry driver.
+        Raises:
+            AquaError: if a solver other than VQE or a variational form other than UCCSD is provided
+                       or if the algorithm finishes due to an unforeseen reason.
+        Returns:
+            A fermionic ground state result.
```

Comment:
More precisely: an `AdaptVQEResult` which provides additional information about how the algorithm ran.

Reply:
Updated by you

Reply:
Resolved and about to push.